Sources/Plasma/PubUtilLib/plPhysX/plPXPhysicalControllerCore.cpp

```diff
@@ -101,17 +118,217 @@ class plPXControllerFilterCallback : public physx::PxQueryFilterCallback
+        plKickableLog_Green(
+            "Applying force ({.2f}, {.2f}, {.2f})",
+            force.x, force.y, force.z
+        );
+        plKickableLog_Green(
+            "Applying torque ({.2f}, {.2f}, {.2f})",
+            torque.x, torque.y, torque.z
+        );
+    }
+
+public:
+    void onShapeHit(const physx::PxControllerShapeHit& hit) override
+    {
+        plPXActorData* controller = static_cast<plPXActorData*>(hit.controller->getUserData());
+        plPXActorData* actor = static_cast<plPXActorData*>(hit.actor->userData);
+        if (controller == nullptr || actor == nullptr)
+            return;
+
+        controller->GetController()->AddContact(
+            actor->GetPhysical(),
+            plPXConvert::ExtPoint(hit.worldPos),
+            plPXConvert::Vector(hit.worldNormal),
+            plPXConvert::Vector(hit.dir),
+            hit.length
+        );
+
+        // Apply some hit force to kickables - the avatar is not actually a simulated body,
+        // so we fake it using some back-of-napkin math here.
+        if (actor->GetPhysical()->IsDynamic()) {
+            physx::PxRigidDynamic* dynamic = static_cast<physx::PxRigidDynamic*>(hit.actor);
+            plKickableLog_White("-----");
+            plKickableLog_Yellow(
+                "Controller '{}' hit dynamic '{}'\nDirection: ({.2f}, {.2f}, {.2f}), Length: {.2f}",
+                controller->str(), actor->str(), hit.dir.x, hit.dir.y, hit.dir.z, hit.length
+            );
+
+            // We need to be careful about when we apply vertical force. In the real world, if you stand
+            // on the ground, you push down on it, and the ground pushes back up on you. In the simulation,
+            // there are no opposite forces, so continued downward force can cause kickables to be killed through
+            // the floor. This is bad news for Eder Gira. But, wait! We would prefer to not be able to ride a
+            // kickable rocket up into the sky because of a log bouncing off the ground. Further complicating matters,
+            // small kickables like said logs will pretty much always be considered hit from amove. So that means
+            //     - Only apply downward forces if we detect the kickable is not already close to the ground.
+            //     - If the kickable is flat-ish, we need a fake force based on the avatar's direction of
+            //       travel; otherwise, it's simply stuck underneath us.
+            physx::PxVec3 hitDir = hit.dir;
+
+            plPXFilterData filterData;
+            filterData.SetGroups((1 << plSimDefs::kGroupStatic) |
+                                 (1 << plSimDefs::kGroupDynamicBlocker));
+            physx::PxQueryFilterData queryFilter(filterData, physx::PxQueryFlag::eSTATIC |
+                                                             physx::PxQueryFlag::eDYNAMIC |
+                                                             physx::PxQueryFlag::ePREFILTER);
```

Comment:
Yes, this can be sort of confusing in that a PhysX dynamc and a Plasma dynamic are not the same thing. A Plasma dynamic is a kickable object. However, in PhysX, a dynamic is any collider that can move, such as a collider controlled by an animation (ie kinematic actors). When we are trying to detect the "ground," we don't want kickables (`plSimDefs::kGroupDynamic`), but we do want the columns in Jalak (`physx::PxQueryFlag::eDYNAMIC`).